ttnn/cpp/ttnn/operations/experimental/cnn/convert_to_hwc/device/kernels/writer_convert_to_hwc.cpp

```diff
@@ -0,0 +1,99 @@
+    constexpr uint32_t cb_in_transpose = get_compile_time_arg_val(1);
+    constexpr uint32_t cb_out = get_compile_time_arg_val(2);
+    constexpr uint32_t channels = get_compile_time_arg_val(3);  // stick size
+    constexpr uint32_t hw = get_compile_time_arg_val(4);        // total number of sticks to copy into output
+    constexpr uint32_t num_full_tiles = get_compile_time_arg_val(5);
+    constexpr uint32_t output_stride_sticks = get_compile_time_arg_val(6);
+    constexpr uint32_t initial_l1_write_stick_offset = get_compile_time_arg_val(7);
+    constexpr uint32_t element_size_bytes = get_compile_time_arg_val(8);
+
+    static_assert(hw % TILE_SIZE == 0, "Shard width must be multiple of tile width");
+
+    constexpr bool is_input_in_dram = get_compile_time_arg_val(9);
+    constexpr bool should_wait = get_compile_time_arg_val(10);
+    constexpr uint32_t dram_write_stride_bytes = get_compile_time_arg_val(11);
+    constexpr uint32_t dram_read_stride_bytes = get_compile_time_arg_val(12);
+    constexpr uint32_t input_sticks_per_core = get_compile_time_arg_val(13);
+
+    if constexpr (is_input_in_dram) {
+        const uint32_t dram_base_read_addr = get_arg_val<uint32_t>(0);
+        const uint32_t num_segments = get_arg_val<uint32_t>(1);
+        tt_l1_ptr uint32_t* args = (tt_l1_ptr uint32_t*)(get_arg_addr(2));
+        uint32_t args_idx = 0;
+        for (uint32_t i = 0; i < num_segments; ++i) {
+            uint32_t copy_size = args[args_idx++];
+            uint32_t write_offset = args[args_idx++];
+            uint32_t bank_id = args[args_idx++];
+            uint32_t read_offset = args[args_idx++];
+            copy_segment_from_dram<dram_read_stride_bytes, dram_write_stride_bytes, input_sticks_per_core>(
+                copy_size, get_write_ptr(cb_in), write_offset, bank_id, dram_base_read_addr, read_offset);
+        }
+        noc_async_read_barrier();
+
+        // One writer must wait until the other has pushed to avoid a race that will trigger a hang
+        if constexpr (should_wait) {
+            cb_wait_front(cb_in, input_sticks_per_core);
+        }
+        cb_push_back(cb_in, input_sticks_per_core);
+    }
+
+    constexpr uint32_t tile_size_stick_bytes = TILE_SIZE * element_size_bytes;
+    constexpr uint32_t channel_size = channels * element_size_bytes;
+
+    constexpr uint32_t l1_write_addr_stride = output_stride_sticks * channel_size;
+    constexpr uint32_t initial_l1_write_addr_offset = initial_l1_write_stick_offset * channel_size;
+
+    const uint32_t base_l1_write_addr = get_write_ptr(cb_out) + initial_l1_write_addr_offset;
+
+    uint32_t l1_write_addr = base_l1_write_addr;
+    for (uint32_t i = 0; i < num_full_tiles; i++) {
+        cb_wait_front(cb_in_transpose, 1);
+        const uint32_t l1_read_addr = get_read_ptr(cb_in_transpose);
+        copy_padded_sticks<channel_size, tile_size_stick_bytes, TILE_SIZE>(l1_read_addr, l1_write_addr);
+        noc_async_read_barrier();
```

Comment:
is there a reason to do a barrier after each copy or we could have one after the loop?